src/platform_linux/shmem.c

```diff
@@ -943,11 +1414,65 @@ platform_shm_find_free(shmem_info *shm,
                   local_in_use);
       }
       platform_shm_trace_allocs(
-         shm, size, msg, retptr, objname, func, file, lineno);
+         shm, size, msg, retptr, objname, func, file, line);
    }
    return retptr;
 }
 
+/*
+ * -----------------------------------------------------------------------------
+ * platform_shm_find_frag() - Find a small free-fragment in a cached list of
+ * free fragments that we track, for specific buckets of fragment sizes.
+ * If one is found of the suitable size, detach it from the list and return
+ * start address of the free fragment. Otherwise, return NULL.
+ *
+ * NOTE: As the free-fragments are linked using free_frag_hdr{}, we return
+ * the address of the recycled free-fragment, which can be temporarily
+ * read as free_frag_hdr{} *.
+ * -----------------------------------------------------------------------------
+ */
+static void *
+platform_shm_find_frag(shmem_info *shm,
+                       size_t      size,
+                       const char *objname,
+                       const char *func,
+                       const char *file,
+                       const int   line)
+{
+   // Currently, we have only implemented tracking small free fragments of
+   // 'known' sizes that appear in our workloads.
+   if ((size < SHM_SMALL_FRAG_MIN_SIZE) || (size > SHM_SMALL_FRAG_MAX_SIZE)) {
+      return NULL;
+   }
+
+   // If we are not tracking fragments of this size, nothing further to do.
+   free_frag_hdr **next_frag = platform_shm_free_frag_hdr(shm, size);
+   if (next_frag == NULL) {
+      return NULL;
+   }
+   shm_lock_mem_frags(shm);
+
+   // Find the next free frag which is big enough
+   while (*next_frag && ((*next_frag)->free_frag_size < size)) {
+      next_frag = &(*next_frag)->free_frag_next;
```

Comment:
Again, could improve this to do a 'best-fit' search.

However, for buckets of size `<= 64` and `<= 128`, by nature of cache line alignments, all free fragments will per force be of 64 or 128 bytes.

The advantages of best-fit or right-fit could be seen only in higher sized buckets: <=256 or higher ...

For another rainy day ...